src/cyw43_bthci_uart.c

```diff
@@ -0,0 +1,326 @@
+/*
+ * This file is part of the cyw43-driver
+ *
+ * Copyright (C) 2019-2025 George Robotics Pty Ltd
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ * 3. Any redistribution, use, or modification in source or binary form is done
+ *    solely for personal benefit and not for any commercial purpose or for
+ *    monetary gain.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE LICENSOR AND COPYRIGHT OWNER "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE LICENSOR OR COPYRIGHT OWNER BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * This software is also available for use with certain devices under different
+ * terms, as set out in the top level LICENSE file.  For commercial licensing
+ * options please email contact@georgerobotics.com.au.
+ */
+
+// CYW43 Bluetooth HCI-over-UART low-level driver.
+
+#include "lib/cyw43-driver/src/cyw43_config.h"
+
+#if CYW43_BT_USE_UART
+
+// This firmware file defines btfw_data and btfw_len.
+#include CYW43_BT_FIRMWARE_INCLUDE_FILE
+
+#define HCI_CMD_HEADER_LEN (4)
+#define HCI_RESP_HEADER_LEN (6)
+#define HCI_PAYLOAD_LEN_MAX (256)
+
+#ifndef CYW43_BT_UART_BAUDRATE_STARTUP
+#define CYW43_BT_UART_BAUDRATE_STARTUP (115200)
+#endif
+
+#ifdef CYW43_PIN_BT_DEV_WAKE
+#define CYW43_BT_SLEEP_TIMEOUT_MS (500)
+static uint32_t cyw43_bluetooth_sleep_ticks;
+#endif
+
+#define CYW43_BT_WAIT_CTS_LOW_TIMEOUT_US (200000)
+#define CYW43_BT_UART_READCHAR_BLOCKING_TIMEOUT_US (100000)
+
+static inline void cyw43_bluetooth_put_le16(uint8_t *buf, uint16_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+}
+
+static inline void cyw43_bluetooth_put_le32(uint8_t *buf, uint32_t val) {
+    buf[0] = val;
+    buf[1] = val >> 8;
+    buf[2] = val >> 16;
+    buf[3] = val >> 24;
+}
```

Comment:
If `CYW43_BT_UART_BAUDRATE_DOWNLOAD_FIRMWARE` and `CYW43_BT_UART_BAUDRATE_ACTIVE_USE` are not defined, I think these static functions will be unused and issue warnings.

Reply:
> maybe not, because they're inline

Right, because they are `static inline` they don't need to be used, and you won't get a warning.

> Also, this is probably an overkill, but if the host byte order can change then you might want to add the following, and handle big endian:

Actually, the reason to write these endian functions explicitly like this is so they are portable across both little and big endian archs.  The point is that the HCI command to change the baudrate is specified to take parameters in little endian order, so that's how the packet is created.

Reply:
clang does indeed give a warning about `static inline` functions not being used.  So I've added the #if guard.

Reply:
I've had similar issues with clang, but `inline` is merely a suggestion, isn't it? So it wouldn't know if it's called or not until later stages I guess.